### PR TITLE
optimize node placement policy in initialized_value stage.

### PR DESCRIPTION
**Pull request by shanshanpt**
Optimize node placement policy in initialized_value stage.
Before optimize, the graph is as follows:
before-optimize
The variable 'global_step' is placed to ps1, and 'global_step/initial_value' is placed to worker. The 'IsVariableInitialized' and 'cond/read/Switch' nodes will be colocated to 'global_step', the 'cond/Switch_1' will be colocated to 'global_step/initial_value'. The left three nodes are not specified to which device, so they will be placed to ps0 by Placer in default. The placement brings some Send/Recv nodes between ps0 and ps1. So I think we can optimize node placement policy in initialized_value stage by colocate these three nodes with the variable(here is global_step). :)

### CODE REVIEW

**Comment by shanshanpt** [comment]
I signed it!

**Comment by alextp** [comment]
This colocate_with was intentionally removed.

The issue is that if the code which computes the initializer has an operation which is not defined for an accelerator device (TPU or GPU) this colocate statement will force the placement of that variable on the CPU, which is a pretty big performance cliff.

Ideally we'd be able to say things like "colocate on the same host, and on the same device if possible" but the placer doesn't have enough knowledge to do this.

My longer-term plan is to wait for placement optimization policies in grappler or xla to get better enough that we can get rid of this.

**Comment by alextp** [comment]
This colocate_with was intentionally removed.

The issue is that if the code which computes the initializer has an operation which is not defined for an accelerator device (TPU or GPU) this colocate statement will force the placement of that variable on the CPU, which is a pretty big performance cliff.

Ideally we'd be able to say things like "colocate on the same host, and on the same device if possible" but the placer doesn't have enough knowledge to do this.

My longer-term plan is to wait for placement optimization policies in grappler or xla to get better enough that we can get rid of this.

**Comment by alextp** [comment]
Sorry for the spammy comments, github was not letting me comment for a while.

File: tensorflow/python/ops/variables.py [hunk 1]
```diff
@@ -511,9 +511,10 @@ def initialized_value(self):
       has run.
     """
     with ops.init_scope():
-      return control_flow_ops.cond(is_variable_initialized(self),
-                                   self.read_value,
-                                   lambda: self.initial_value)
+      with ops.colocate_with(self):
```
[the remark this answers is not in the document]
Comment (shanshanpt) [reply tensorflow/python/ops/variables.py]: @alextp Thanks for your reply.

**Pull request closed without merging**